docs/proposals/15-manage-bastions-and-ssh-key-pair-rotation.md

```diff
@@ -0,0 +1,212 @@
+  - Similar to `BackupBucket`s or `BackupEntry`, the `gardenlet` watches the `Bastion` resource in the garden cluster and creates a seed-local `Bastion` resource, on which the provider specific bastion controller acts upon
+- `gardenctlv2` (or any other client)
+  - Creates `Bastion` resource in the garden cluster
+  - Establishes an `ssh` connection to a shoot node, using a bastion host as proxy
+  - Heartbeats / keeps alive the `Bastion` resource during `ssh` connection
+- Gardener extension provider <infra>
+  - Provider specific bastion controller
+  - Should be added to gardener-extension-provider-<infra> repos, e.g. https://github.com/gardener/gardener-extension-provider-aws/tree/master/pkg/controller
+  - Has the permission to update the `Bastion/status` subresource on the seed cluster
+  - Runs on seed (of course)
+- Gardener Controller Manager (`GCM`)
+  - `Bastion` heartbeat controller
+    - Cleans up `Bastion` resource on missing heartbeat.
+    - Is configured with a `maxLifetime` and `timeToLife` for the `Bastion` resource
+- Gardener (RBAC)
+  - The project `admin` role should be extended to allow CRUD operations on the `Bastion` resource. The `gardener.cloud:system:project-member-aggregation` `ClusterRole` needs to be updated accordingly (https://github.com/gardener/gardener/blob/master/charts/gardener/controlplane/charts/application/templates/rbac-user.yaml)
+
+### SSH Flow
+0. Users should only get the RBAC permission to `create` / `update` `Bastion` resources for a namespace, if they should be allowed to `ssh` onto the shoot nodes in this namespace. A project member with `admin` role will have these permissions.
+1. User/`gardenctlv2` creates `Bastion` resource in garden cluster (see [resource example](#resource-example) below)
+    - First, gardenctl would figure out the own public IP of the user's machine. Either by calling an external service (gardenctl (v1) uses https://github.com/gardener/gardenctl/blob/master/pkg/cmd/miscellaneous.go#L226) or by calling a binary that prints the public IP(s) to stdout. The binary should be configurable. The result is set under `spec.ingress[].ipBlock.cidr`
+    - the public `ssh` key of the user is set under `spec.sshPublicKey`. The key needs to be configured beforehand by the user
+    - The targeted shoot is set under `spec.shootRef`
+2. GAPI Admission Plugin for the `Bastion` resource in the garden cluster
+        - on creation, sets `metadata.annotations["operations.gardener.cloud/created-by"]` according to the user that created the resource
+        - when `operations.gardener.cloud/operation: keepalive` is set it will be removed by GAPI from the annotations and `status.lastHeartbeatTimestamp` will be set with the current timestamp. The `status.expirationTimestamp` will be calculated by taking the last heartbeat timestamp and adding x minutes (configurable, default `60` Minutes).
```

Comment:
Can we do the heartbeats via `leases`? This is the proper resource to use for such requirements, so we should use it instead of re-implementing it via own mechanisms.

Reply:
I also wondered if I should use the `Lease` API but I decided against it as I think it adds unneeded complexity.

The goal for introducing the `Lease` API was

> ### Goals
> 
> - Reduce size of etcd by making node heartbeats cheaper
> 
> ### Non-Goals
> 
> The following are nice-to-haves, but not primary goals:
> 
> - Reduce resource usage (cpu/memory) of control plane (e.g. due to processing
>   less and/or smaller objects)
> - Reduce watch-related load on Node objects

https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/589-efficient-node-heartbeats/README.md

As `gardenctl` would run the bastion heartbeats every `x` minutes (not seconds); plus I expect the `Bastion` resources to not be that large, I don't think it would be that beneficial to use the `Lease` API.
Also we can better control who (only the creator of the `Bastion` resource) can keep the bastion alive.

Reply:
Ok, maybe you're right. Then, let's go with that proposal first and change it later if we encounter issues with it.